### PR TITLE
Angela - MailChimp Replacement Backend

### PR DESCRIPTION
**Pull request by angelalalacheng**
# Description
- We currently use MailChimp to send out our weekly progress updates. It is not user friendly and the price keeps going up. I’d like a replacement that an Admin can use via the app.
- Take over the work of previous people : [FE](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1315 ) [BE](https://github.com/OneCommunityGlobal/HGNRest/pull/540 ) led to https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1784
- [Video of what’s needed](https://www.loom.com/share/ce0ff3e66f3e40e19703ad9f4206cad1?sid=775e17db-d792-4a1c-a6d6-26379ccfe48c)

## Related PRS (if any):
To test this backend PR you need to checkout [#2595 frontend PR](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2595).

## Main changes explained:
- Update `src/controllers/emailController.js` for refactoring and use content id to ensure that the image can display correctly in Gmail
- Update `src/utilities/emailSender.js` for refactoring the code

## Before Testing
To send an email using the Gmail API with OAuth2 and nodemailer, follow the video below and set it up.
https://www.youtube.com/watch?v=-rcRf7yswfM&t=10s&ab_channel=MafiaCodes

Fill in the these variable values from the above setup in the .env file of the backend.
REACT_APP_EMAIL =
REACT_APP_EMAIL_CLIENT_ID =
REACT_APP_EMAIL_CLIENT_SECRET =
REACT_APP_EMAIL_REFRESH_TOKEN =
REACT_APP_EMAIL_CLIENT_REDIRECT_URI = https://developers.google.com/oauthplayground
sendEmail = true

## How to test:
1. check into current branch
2. do `npm install` and `npm run dev` to run this PR locally
3. Clear site data/cache
4. If you only want to send the testing email to your own testing accounts, you can add a search keyword in src/controllers/emailController.js like the following:
Screenshot 2024-08-20 at 9 23 17 PM

5. See how to test in frontend [#2595 frontend PR](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2595)

## Screenshots or videos of changes:
See frontend

## Note:
See frontend

### CODE REVIEW

**Review by vishnupriyaatheti — COMMENTED** [review]
@angelalalacheng I am trying to setup  that was described in youtube video. To authorize the Api I am getting below attached error. Can you please suggest what wrong I am doing.
BE PR 1080

**Comment by angelalalacheng** [comment]
> @angelalalacheng I am trying to setup that was described in youtube video. To authorize the Api I am getting below attached error. Can you please suggest what wrong I am doing. BE PR 1080

Hi, @vishnupriyaatheti 
Sorry about that I have no idea about that. I google it and also asked GPT andbutnot sure whether it is helpful.

The issue you’re facing is due to the fact that your application hasn’t completed the Google API verification process. When using Google APIs, such as the Gmail API, the app needs to go through a security and verification process by Google to ensure it complies with privacy and security standards.

The error message “Access blocked: Gmail API tutorial has not completed the Google verification process” indicates that the app is still in testing and can only be accessed by developer-approved testers. The app is blocked from wider access because it hasn’t been fully verified.

**Review by Ankuriboh — CHANGES_REQUESTED** [review]
Code was great, except that we need to reconsider the necessity of the queue system.

Asides from this, it somewhat a hassle to use Gmail as our email provider. I would suggest that we move to a real bulk mail sender other than Mailchimp to solve the issue.

File: src/utilities/emailSender.js [hunk 1]
```diff
@@ -2,85 +2,120 @@ const nodemailer = require('nodemailer');
...
-      // Why?
-      // 1. Could create a security risk
-      // 2. Could create heavy loads on the server if emails are sent to many people
-      // 3. Contain limited useful info:
-      //   result format : {"accepted":["emailAddr"],"rejected":[],"envelopeTime":209,"messageTime":566,"messageSize":317,"response":"250 2.0.0 OK  17***69 p11-2***322qvd.85 - gsmtp","envelope":{"from":"emailAddr", "to":"emailAddr"}}
-      if (process.env.NODE_ENV === 'local') {
-        logger.logInfo(`Email sent: ${JSON.stringify(result)}`);
+    };
+
+    const processBatch = async (batch) => {
```
[the remark this answers is not in the document]
Comment (angelalalacheng) [reply src/utilities/emailSender.js:86]: Thank you for the suggestion! The code extends from previous developer so I just keep use it but I can modify based on your comment!

**Review by ajitp4292 — COMMENTED** [review]
Awaiting the Google verification to Complete.
Screenshot 2024-11-09 at 6 15 31 PM

**Review by one-community — APPROVED** [review]

**Pull request merged**